### PR TITLE
Filter out channel join messages for topchannels

### PR DESCRIPTION
**Pull request by ajkerrigan**
Allow karmabot to ignore certain message subtypes for the purposes of `topchannels` reporting. If the latest message in a channel is of an ignored subtype (initially only `channel_join`), look through channel history for the last message that would not be ignored. Channels with _only_ ignored messages are excluded from topchannels reports.

The logic of this seems to be finished, but I need to flesh out the mocks and tests. Putting it up for eyeball reviews and comments in the meantime.

### CODE REVIEW

File: commands/topchannels.py [hunk 1]
```diff
@@ -97,16 +107,54 @@ def calc_channel_score(channel: Channel):
...
+        if not grant_user_token:
+            logging.info(
+                "Cannot search channel history, no env SLACK_KARMA_INVITE_USER_TOKEN"
+            )
+            return None
+
+        sc = SlackClient(grant_user_token)
+        response = sc.api_call("channels.history", channel=channel.id, user=karmabot_id)
+
+        msgs = [
```
Comment (bbelderbos) [inline commands/topchannels.py]: you could return msgs, if it's empty it's "falsy"

File: commands/topchannels.py [hunk 2]
```diff
@@ -97,16 +107,54 @@ def calc_channel_score(channel: Channel):
...
+            msg
+            for msg in response["messages"]
+            if msg.get("subtype") not in ignore_message_types
+        ]
+        if not msgs:
+            return None
+        latest_ts = float(max(msgs, key=itemgetter("ts"))["ts"])
```
Comment (bbelderbos) [inline commands/topchannels.py]: `itemgetter` is neat!

File: commands/topchannels.py [hunk 3]
```diff
@@ -79,16 +83,22 @@ def get_recommended_channels(**kwargs):
...
     potential_channels = sorted(
-        potential_channels, key=calc_channel_score, reverse=True
+        ((calc_channel_score(chan), chan) for chan in potential_channels), reverse=True
     )
 
-    for channel in potential_channels[:nr_channels]:
-        msg += f'{LIST_ICON} #{channel.name} ({channel.num_members} members, {seconds_since_last_post(channel) / 3600:.0f} hour(s) since last post): {channel.purpose if channel.purpose else "<Invest today and get an awesome description!>"}\n'
+    msg = MSG_BEGIN + "\n".join(
+        (
+            f'{LIST_ICON} #{channel.name} ({channel.num_members} members, {seconds_since_last_post(channel) / 3600:.0f} hour(s) since last post): {channel.purpose if channel.purpose else "<Invest today and get an awesome description!>"}'
```
Comment (bbelderbos) [inline commands/topchannels.py]: not very readable, I would make a template string constant at the top and `format` the variables in here

Comment (ajkerrigan) [reply commands/topchannels.py]: Good call 👍

File: commands/topchannels.py (hunk 1 above)
Comment (ajkerrigan) [reply commands/topchannels.py]: Hmm I'm torn on that. The method returns a float. Making it `Optional` to allow for a `None` return seems logical enough, but having it return `float` or a falsy empty list feels awkward.

Maybe it's worth a refactor so we don't call `seconds_since_last_post` at all unless we have a valid (not ignored) message timestamp.

Comment (pybites) [reply commands/topchannels.py]: Good point, thanks

**Comment by pybites** [comment]
Thanks a lot @ajkerrigan will deploy this.

**Pull request merged**